File: adt2sys/go32/pas/dump_VESA_buffer.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure dump_VESA_buffer(buffer_size: Longint);

var
  dumped_data_size,bank_data_size: Longint;
  current_bank: Byte;

begin
  If NOT _draw_screen_without_vsync then
    WaitRetrace;
  dumped_data_size := 0;
  current_bank := 0;
  While (dumped_data_size < buffer_size) do
    begin
      If (dumped_data_size+65536 <= buffer_size) then
        bank_data_size := 65536
      else bank_data_size := buffer_size-dumped_data_size;
      VESA_SwitchBank(current_bank);
      dosmemput($0a000,0,_FrameBuffer_mirror[dumped_data_size],bank_data_size);
      Inc(dumped_data_size,bank_data_size);
      Inc(current_bank);
    end;
end;

File: adt2sys/go32/dump_VESA_buffer.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void dump_VESA_buffer (ssize_t buffer_size) {
  ssize_t left_sz = buffer_size;
  const uint8_t *p = _FrameBuffer_mirror;
  uint8_t bank = 0;

  if (!_draw_screen_without_vsync)
    WaitRetrace ();

  while (left_sz > 0) {
    ssize_t copy_sz = (left_sz > 0x10000) ? 0x10000 : left_sz;

    VESA_SwitchBank (bank);
    dosmemput (p, copy_sz, 0xA0000);
    left_sz -= copy_sz;
    p += copy_sz;
    bank++;
  }
}
